### ISSUE TITLE
Multiple connections for boards with multiple ethernet connected chips

### ISSUE THREAD

**Issue by mundya** [project-rig/rig#142]
Not immediately required (it would be nice to have for Canada, possibly) but definitely necessary for Spaun.

As far as I can tell, the IP address of a chip can be read from the `sv` struct (`ip_addr`) and is, I assume formatted the same way as IP addresses for IP tags (e.g., `!4B`, bewaring that the default appears to be 255.255.0.0).
The real challenge is in determining the co-ordinates of likely chips, though I guess a safe approach would be to always jump 8* (north or east) from the last found ethernet chip and then read `eth_addr` to get the co-ordinates of the nearest ethernet connected chip.

I initially thought that this process should be part of `get_machine`... but now I'm not so sure.

Once we have a map of co-ordinates to ethernet addresses it becomes a drop in replacement to `MachineController.get_connection`, though we should also expose the mapping so that other software (like Nengo) can transmit SDP to SpiNNaker efficiently.

\* or 7?

Also: "bewaring" is an awesome word.

**Comment by mossblaser** [comment]
There is of course the issue of how you parallelise all your reads/writes (i.e. using multiple Ethernets at once) with a nice-ish API... Asyncio-type approaches seem to be the main candidate here but slooowww!

That aside, you may find the following functions from the defunct asyncio branch useful for getting the set of candidate ethernet-connected chips:

https://github.com/project-rig/rig/blob/b666e802f97a7321c3c8a1192495744949e0c5b5/rig/geometry.py#L176-L254

When I was playing with this, I built the following machine discovery implementation:

https://github.com/project-rig/rig/blob/b666e802f97a7321c3c8a1192495744949e0c5b5/rig/machine_control/machine_controller.py#L231-L327

If you're worrying about IP address and ethernet connectivity stuff, this code should have you covered. Note the `sv->eth_up` field for detecting up-ness and that the least-significant-eight-bits of the IP returned are the first field in a conventionally rendered IP address.

**Comment by mundya** [comment]
Awesome, thank you!

> There is of course the issue of how you parallelise all your reads/writes (i.e. using multiple Ethernets at once) with a nice-ish API... Asyncio-type approaches seem to be the main candidate here but slooowww!

On my "to test" list is making `SCPConnection` thread safe and seeing what hellish results come from writing in two threads at once.

**Comment by mossblaser** [comment]
Following from a recent offline discussion with @mundya about speeding up I/O for large machines using multiple Ethernet connections, the following conclusions were reached:

* As ever, in principle, the only operation worth speeding up is read/write since all other operations are broadcast (and thus already fast enough) or for diagnostics only (and thus for interactive-style use and so already fast enough).
* The "hard-bit" is extending the existing file-like metaphor without requiring apps to be reorganised. Using AsyncIO is probably not a workable option.

Possible options:

* For writing:
  * The buffering mechanism can potentially be extended to allow a global "flush()" call to be made which uses as many Ethernet connections as available. Disadvantage is we end up having everything we want to load placed in memory at once.
  * To work-around the above disadvantage, we could add a call-back-on-write option which allows writes to be defined by an address, a length and a callback which is called when the data is actually required. This complicates life somewhat, of course.
* For reading:
  * Conventional read interfaces clearly aren't going to cut it.
  * Could do call-back-on-read or read-into-buffer; obviously one is messier, the other requires everything being in memory at once and sounds just a little bit dangerous too.

Clearly all this requires some new centralised state:

* Queues of outstanding reads/writes (one per chip or per connection) and what to do with them

This is going to prove faffy to get right...

**Comment by mossblaser** [comment]
The following is an informal design document for multi-connection support. Feedback highly encouraged.

High-Throughput IO Plans
========================

Getting data in and out of SpiNNaker machines with high throughput is very
important for large multi-board systems with potentially around six gigabytes
of data per board to be loaded and only a single fairly measly 100MBit/s
interface each. Rig's new SCP I/O implementation with windowing enables the
best possible throughput into a single Ethernet connection. In this PR I hope
to extend this to allowing multiple connections to several Ethernet connected
chips to be used. It is not anticipated that anything but bulk memory
read/writes will be required to perform high throughput I/O since all other
operations are either broadcast-like or diagnostic (and thus fast enough at the
moment).

How many connections?
---------------------

Though ideally Rig should be able to drive all the Ethernet connections to an
arbitrarily large machine at once, in practice this is simply infeasible for
the following reasons:

* The Python SCP implementation is likely to become CPU bound once a sufficient
  number of SCP connections are in use since individual SCP packets are very
  small.
* A single Ethernet connection is more-or-less going to offer 1 GBit/s of I/O
  which is only enough to saturate 10 boards' Ethernet connections anyway
  (though in practice board Ethernet bandwidth is severely constrained anyway).

I'm not sure which of these two issues will strike first. Initially I'll hope
for the best and assume we get I/O bound before CPU bound.

Which connection should be used?
--------------------------------

I would suggest that loading is always performed via the Ethernet connection on
the same board as the target processor. This choice has the advantage of being
easily resolved via a lookup table. When a connection is not present, two
options exist:

1. Fall back on (0, 0)
2. Fall back on the nearest working connection
3. Fall back on a randomly chosen connection
4. Fall back on a cyclically chosen connection

Option 1 is very easy but may lead to significant load imbalance: in principle
N dead Ethernet connections could mean N times longer load times.

Option 2 in principle is ideal but in practice is expensive since it requires
a fancy spatial lookup function to discover. A lookup table could be used but
this seems wasteful...

Option 3 and 4 potentially work reasonably well and probably differ mainly in
expense of computation. It probably isn't worth worrying about the difference
in cost...

Overall, option 3 is probably the easiest to implement and stands the best
chance of working.

How should multiple connections be used?
----------------------------------------

In principle it could be done with a single select statement and some clever
logic. Alternatively, we can run many instances of the current SCPConnection
objects in separate threads. If we're truly I/O bound these two approaches
differ only in the relative costs of Python vs OS level dispatching
performance.

If this were C, the select option would be the way to go but since this is
Python there's a decent chance the difference in performance could be
comparable. The threading option is also potentially much easier to implement
so I'll be going with that. If the SCP sending code proves too CPU intensive
(sigh...) multiprocessing may be an option if there turns out to be some
moderately efficient way to share memory.

I propose having a thread pool with each thread allocated a particular SCP
connection. When more SCP connections are present than threads, the remaining
connections (and their associated blocks of data) should be queued and given to
threads which have exhausted the traffic associated with their connection.

As for how to select the number of threads; initially this can just be fixed
but eventually some kind of control system could be set up...

How should data be supplied?
----------------------------

Two obvious possibilities exist:

1. Data/free memory could be supplied in full up-front.
2. Callbacks could be arranged which return/handle data as it is fetched.

The former is potentially the easiest to implement and has the cleanest API.
Unfortunately it may require your PC to have vast amounts of memory for very
large problems. Further, there is lots of potential for subtle bugs arising
from accidentally using the same memory twice or accessing uninitialised memory
buffers.

The later is much more fiddly to implement but obviously is more scalable since
clever users can process their data as it arrives. It would be possible to wrap
this approach to look like the first approach for ease of implementation. This
approach may also be marginally more costly due to the added function call
overhead.

I think due to safety and scalability I'll regrettably have to go with the
second option as first class citizen and the first as an alternative API.

How should errors be treated?
-----------------------------

For a large portion of applications, a single error during read/write is
non-recoverable. This is a semi-easy case to deal with, made super-easy if any
data received so far can be safely ignored, e.g. if all that will happen is an
error message and an exit.

Sadly, for larger-scale applications, a partial failure may not be as bad as
complete failure. For example, when reading back the results of an experiment,
if a single core has died during the experiment, the results on other cores may
still be valid and worth downloading. In such a circumstance it is still
important to be able to determine *what* failed afterwards.

The correct solution here is probably to have two alternative modes of
operation.

How should the API look?
------------------------

Files on Linux are famously poor at dealing with multiple asynchronous
reads/writes. On platforms where this is supported it is largely via
non-orthodox API. However this is done, it is probably going to result in the
File-like API being extended beyond typical file-like behaviour.

AsyncIO/Trollius could be used here however it has famously dreadful
performance and even more dreadful documentation. This type of abstraction also
has a tendency to leak out into applications where it is unwanted.

An idea:

    # Thin-wrapper around the current interface; overrides mc.read and mc.write
    # such that they queue up reads/writes. For reads they also return a new
    # buffer with garbage in it.
    mc.begin_burst()
    
    mc.write(...)
    data0 = mc.read(...)
    
    file_like.write(...)
    file_like.write(...)
    file_like.flush()
    
    data1 = file_like.read()
    
    # All the above actions actually take place now and in parallel. If
    # something goes wrong, you get an exception and the above actions
    # may/may-not have been called.
    mc.end_burst()

The above works nicely for upgrading existing applications where you can afford
the memory overhead of having everything in RAM at once. To extend the APIs
somewhat we could make:

* write() accept a callable and length in place of some bytes.
* read() accept a callable as an extra argument and not return

With this modified API, callback style usage could be achieved at the expense
of the file-like API being no-longer file-like. To properly support this, the
write buffering/coalescing behaviour of MemoryIO will need modifying to support
coalescing callbacks. Outside burst-mode, the callbacks will simply get called
immediately.

By default, if an error occurs (including in a callback function) the end_burst
method will simply propagate that error stopping all other activity as cleanly
and quickly as possible. Some callbacks may not be called and some read data
may not have been placed in the reserved buffers (leaving their contents
undefined).

Alternatively, end_burst could accumulate exceptions and tracebacks, continuing
as far as possible, and return a list of (callback, exception, traceback)
tuples. Obviously, the callback object won't make sense unless you're using
your own callback system but given that this is a sufficiently advanced mode of
operation, this is probably what should be used anyway.

What will happen to the internals?
----------------------------------

0. The begin_burst function will simply set a flag.
1. A set of read_into() methods will need to be added to allow creation of the
   read buffer in advance.
2. For each connection there will be a queue of read/write requests.
3. All read/write requests should be wrapped in a lambda(?) and placed in the
   buffer. For conventional reads, a buffer should be created and then
   read_into called with that buffer.
4. When not in a burst, the buffer will be emptied immediately (or a fast path
   used...).
5. The end_burst function will spawn some threads and in each thread work on
   processing the queues.

**Comment by mossblaser** [comment]
An alternative API idea:

* MachineController.read/write could accept an argument which optionally postpones the read/write.
* When MemoryIOs are created, the above argument can be chosen which then applies to all uses of that memoryio.

This alternative has the following advantages:

* Internal use of read/write (e.g. reading struct variables) don't need to be modified to assert that burst mode is not being used
* It makes the whole thing a lot more explicit without too much hassle (and can even be enabled via the context manager!). This is good since things like reads producing garbage are a ripe source of bugs for beginners!
* Reduces the amount of "leaky" state (added state is now reduced to just the set of queues)

**Comment by mundya** [comment]
> An alternative API idea:
> 
>  - MachineController.read/write could accept an argument which optionally postpones the read/write.
>  - When MemoryIOs are created, the above argument can be chosen which then applies to all uses of that memoryio.

I like this, with the modification that the additional argument to `read` and `write` is a callback function. If the function is `None` (the default) then the read/write is blocking, otherwise the read/write is asynchronous and the callback is called on completion with the read data/number of bytes written. `MemoryIO` can just expose these arguments.

We can add `flush`, or similar, to the machine controller to block until all queued events are processed.

Thoughts? The advantages are the same as above but with fewer API changes (and with complete backwards compatibility).

**Comment by mossblaser** [comment]
What would your use-case for the callback function be? My understanding of most loads is that a sync/flush function is the simplest thing which could possibly work and should result in minimal changes to existing code. Certainly much simpler than callbacks etc. for this style of application.

My reluctance to exposing a callback is that the "write and flush/sync later " API does not imply any sort of asynchronous programming mechanism while a callback does and I don't have a good feeling about that given Python's lack of a good defacto asynchronous programming mechanism.

**Comment by mundya** [comment]
Which callback? The read callback I can see being particularly handy (over, say `read_into` and `sync`, where I have to suspend a load of operations until the `sync` completes).

The write callback was mostly for symmetry over just saying `write(..., async=True)`. I suspect that in most cases the write callback probably would be a NOP but some particularly pedantic people might want to check that they did actually write as much as they intended.

**Comment by mundya** [comment]
[I guess `write(..., immediate=False)` might be nearer the truth...]

**Comment by mossblaser** [comment]
> The read callback I can see being particularly handy (over, say read_into and sync, where I have to suspend a load of operations until the sync completes).

Indeed, but if we provide the callback we are introducing an implied concurrency model to the user's application. This is not a step I'd take too lightly and as mentioned there is no clear "good solution" here. I can't think of an application where you'd want to be notified when some (but not all) of your reads have completed -- what sort of thing are you thinking of?

> The write callback was mostly for symmetry over just saying write(..., async=True). I suspect that in most cases the write callback probably would be a NOP but some particularly pedantic people might want to check that they did actually write as much as they intended.

Fair, though my plan was to have exceptions in read/writes to pop out of the sync function to prevent silent errors in most apps. Having to write a NOP function is also vaguely irritating ;).
